Add Trino 427 release notes (#19023)
trinodb/trino: docs/src/main/sphinx/release/release-427.md

# Release 427 (25 Sep 2023)

## General

> * Add support for IPv4 mapped addresses in `contains`. ({issue}`18497`)

Reviewer: why "mapped" ... ? Also maybe .. `for comparisons with the contains function`

> * Add support for comparing IPv4 addresses and IPv6 CIDRs with the `contains`
>   function. ({issue}`18497`)

Reviewer 2: ```suggestion
* Add support for comparing IPv4 and IPv6 addresses and CIDRs with [contains](ip-address-contains). ({issue}`18497`)
```

> * Improve performance of `GROUP BY`, `DISTINCT`, and other aggregations. ({issue}`19059`)

Reviewer 2: ```suggestion
* Improve performance of `GROUP BY` and `DISTINCT`. ({issue}`19059`)
```

> * Improve performance of `GROUP BY`, `DISTINCT`, and other hashing operations. ({issue}`19059`)

Reviewer: Thats a big one! Do we need to rephrase to something like

Improve performance of hashing operations used for `GROUP BY`, `DISTINCT, and other aggregations.

* Fix query failure due to exceeding node memory limits with `INSERT`
  statements. ({issue}`18771`)

> * Fix incorrectly reporting `ns` instead of `ms` in query plans. ({issue}`19047`)

Reviewer: Maybe 

nanoseconds (ns) instead of milliseconds (ms)

## Security

> * Ensure access control is checked when accessing table comments. ({issue}`18514`)

Reviewer 2: ```suggestion
* Ensure authorization is checked when accessing table comments. ({issue}`18514`)
```

## Delta Lake connector

* Add support for reading tables with

>   [deletion vectors](https://docs.delta.io/latest/delta-deletion-vectors.html). ({issue}`16903`)

Reviewer 2: ```suggestion
  [Deletion Vectors](https://docs.delta.io/latest/delta-deletion-vectors.html). ({issue}`16903`)
```

Author: I don't think this is correct - they're lowercase in all Delta Lake documentation.

Reviewer 3: Both are fine. They use "Deletion Vectors" in their protocol page. https://github.com/delta-io/delta/blob/master/PROTOCOL.md#deletion-vectors

> * Add support for writer

Reviewer: ```suggestion
* Add support for Delta Lake writer version 7
```

Or 

```suggestion
* Add support for Delta Lake protocol version 7
```

Reviewer 3: "Delta Lake writer version 7" is better. We don't call it "protocol version".

  [version 7](https://docs.delta.io/latest/versioning.html#features-by-protocol-version). ({issue}`15873`)

> * Add support for writing the `timestamp` type. ({issue}`16927`)

Reviewer: Only writing? Dont we word this differently normally?

Reviewer: Add support for writing columns with timestamp type 

maybe

Reviewer 3: Only writing. Read is already supported. o

* Reduce data read from Parquet files for queries with filters. ({issue}`19032`)
* Improve performance of writing to Parquet files. ({issue}`19122`)
* Fix error reading Delta Lake table history when the initial transaction logs
  have been removed. ({issue}`18845`)

## Elasticsearch connector

* Fix query failure when a `LIKE` clause contains multi-byte characters. ({issue}`18966`)

## Hive connector

> * Add support for changing column comments in the Glue catalog. ({issue}`19076`)

Reviewer 2: ```suggestion
* Add support for changing column comments when using the Glue catalog. ({issue}`19076`)
```

> * Add automatic type coercion from `varchar(0)` to `varchar(1)`. ({issue}`18811`)

Reviewer: what. . varchar(0) is valid? That is weird

Author: Causes errors according to the PR adding it.

> * Add support for altering column comments in the Glue catalog. ({issue}`19076`)

Reviewer: ```suggestion
* Add support for `ALTER TABLE ... ALTER COLUMN` with the Glue catalog. ({issue}`19076`)
```

Author: This would be `ALTER TABLE ... ALTER COLUMN ... COMMENT "new comment"`, which is too much syntax and loses the clarity to saying it in plain speech, IMO.

Reviewer 3: The syntax is wrong. It should be `COMMENT ON COLUMN`. https://trino.io/docs/current/sql/comment.html

Reviewer: So thats is not about altering the comment .. but about setting it.. we should use the syntax from @ebyhr then

Reviewer 3: `COMMENT ON COLUMN` statement can be used for both adding a new comment and altering the existing comment. 

The current supported operations of `ALTER COLUMN` is only changing column types.

  `*_native_writer_enabled` session properties and `*.native-reader.enabled` and
  `*.native-writer.enabled` configuration properties are removed. ({issue}`18241`)
* Remove support for S3 Select. The `s3_select_pushdown_enabled` session
  property and the `hive.s3select*` configuration properties are removed. ({issue}`18241`)
* Remove support for disabling optimized symlink listing. The
  `optimize_symlink_listing` session property and
  `hive.optimize-symlink-listing` configuration property are removed. ({issue}`18241`)
* Fix incompatibility with Hive OpenCSV deserialization. As a result, when the
  escape character is explicitly set to `"`, a `\` (backslash) must be used
  instead. ({issue}`18918`)
* Fix performance regression when reading CSV files on AWS S3. ({issue}`18976`)

## Hudi connector

* Fix query failure when reading from Hudi tables with
  [`instants`](https://hudi.apache.org/docs/concepts/#timeline) that have been
  replaced. ({issue}`18213`)

## Iceberg connector

* Add support for usage of `date` and `timestamp` arguments in `FOR TIMESTAMP AS
  OF` expressions. ({issue}`14214`)
* Add support for using tags with `AS OF VERSION` queries. ({issue}`19111`)

> * Reduce coordinator memory usage when running queries on tables with large

Reviewer: large metadata .. that sounds weird .. is there some better term? a lot of metadata .. is also weird

  metadata. ({issue}`19009`)
* Reduce data read from Parquet files for queries with filters. ({issue}`19032`)
* Improve performance of writing to Parquet files. ({issue}`19090`)

## Ignite connector

> * Add support for pushing down `UPDATE` statements to Ignite. ({issue}`16445`)

Reviewer: pushdown of UPDATE statements ??

# Release 427 (25 Sep 2023)

## General

> * Add support for IPv4 addresses in comparisons using the `contains` functions. ({issue}`18497`)

Reviewer 2: IPv4 is already supported. This change adds support for comparing IPv4 addresses against IPv6 CIDRs and vice-versa.

> * Make plans with complex aggregations deterministic. ({issue}`19103`)

Reviewer 2: No need to document this. Also, it's incorrect. Plans are already deterministic -- this just makes how they are *printed* more predictable.

> * Improve performance of hashing operations used by `GROUP BY`, `DISTINCT`, and

Reviewer 2: Improve performance of `GROUP BY`, ... (no need to say "hashing operations". That's a low-level detail)

  other aggregations. ({issue}`19059`)

> * Fix query failure due to exceeding node memory limits with `INSERT`

Reviewer 2: Fix failure ... (remove "query")

  statements. ({issue}`18771`)

> * Fix incorrectly reporting nanoseconds instead of milliseconds in query plans. ({issue}`19047`)

Reviewer 2: This is inaccurate. It's when reporting errors due to timeout during planning. No need to document this.

* Fix query hang for certain `LIKE` patterns involving a mix of `%` and `_`. ({issue}`19146`)

## Security

> * Add access control check to redirected tables when accessing table comments. ({issue}`18514`)

Reviewer 2: What's the user impact of this? A "redirected table" is a low-level connector concept, so it's not meaningful to describe it as such.

## Delta Lake connector

> * Add support for reading tables with deletion vectors. ({issue}`16903`)

Reviewer 2: Deletion Vectors

* Add support for Delta Lake writer
  [version 7](https://docs.delta.io/latest/versioning.html#features-by-protocol-version). ({issue}`15873`)

> * Add support for writing columns with the `timestamp` type. ({issue}`16927`)

Reviewer 2: Is this Trino's timestamp type? If so, it should be described as `timestamp(p)` or `timestamp(p) without time zone`

Reviewer 3: Yes, it's Trino timestamp type.

* Reduce data read from Parquet files for queries with filters. ({issue}`19032`)
* Improve performance of writing to Parquet files. ({issue}`19122`)
* Fix error reading Delta Lake table history when the initial transaction logs
  have been removed. ({issue}`18845`)

## Elasticsearch connector

* Fix query failure when a `LIKE` clause contains multi-byte characters. ({issue}`18966`)

## Hive connector

> * Add support for `double` to `varchar` coercion in Hive tables. ({issue}`18930`)

Reviewer 2: This is incorrect. It's not about supporting coercions, but allowing certain kinds of table/column evolutions (e.g., changing column type from `double` to `varchar(n)`)

> * Add automatic type coercion from `varchar(0)` to `varchar(1)`. ({issue}`18811`)

Reviewer 2: This is also misleading. This is about mapping Trino's `varchar(0)` type to `varchar(1)` when creating a table in Hive instead of failing due to the type being unsupported. E.g., when doing:

```
CREATE TABLE t(x) AS (SELECT CAST('' AS varchar(0)));
```

```
trino:tpch> describe t;
 Column |    Type    | Extra | Comment
--------+------------+-------+---------
 x      | varchar(1) |       |
(1 row)
````

# Release 427 (13 Sep 2023)

## Security

> * Added access control check to redirected tables when accessing table comments. ({issue}`18514`)

Reviewer 3: nit: Added → Add

## ClickHouse connector

* Add support for pushing down `UPDATE` statements to ClickHouse. ({issue}`16445`)

## Delta Lake connector

* Add support for reading tables with deletion vectors. ({issue}`16903`)

## Druid connector

* Add support for pushing down `UPDATE` statements to Druid. ({issue}`16445`)

## Hive connector

> * Add support for pushing down `UPDATE` statements to Hive. ({issue}`16445`)

Reviewer: To Hive ... Hive is no running when using Trino .. where are we pushing to .. the HMS ?

Reviewer 4: https://github.com/trinodb/trino/pull/16445 - there is no support for Hive in this PR only JDBC connectors:
Ignite
MariaDB
MySql
Oracle
Postgres
Redshift
Singlestore
SqlServer

* Add support for `double` to `varchar` coercion in Hive tables. ({issue}`18930`)
* Remove legacy Hive readers and writers. The `*_native_reader_enabled` and
  `*_native_writer_enabled` session properties and `*.native-reader.enabled` and

>   `*.native-writer.enabled` configuration properties have been removed. ({issue}`18241`)

Reviewer: I assume docs are updated

Reviewer: Also .. "are removed" .. since the rest is present tense ..

> * Remove support for S3 Select. The `s3_select_pushdown_enabled` session

Reviewer: I assume docs are updated

  property and the `hive.s3select*` configuration properties have been removed. ({issue}`18241`)
* Remove support for disabling optimized symlink listing. The
  `optimize_symlink_listing` session property and

>   `hive.optimize-symlink-listing` configuration property have been removed. ({issue}`18241`)

Reviewer: are remove ?

* Fix incompatibility with Hive OpenCSV deserialization. As a result, when the

>   escape character is explicitly set to `"`, a backslash must be used instead. ({issue}`18918`)

Reviewer: maybe
```
`\` (backslash)
```

> * Fix slow reading of CSV files on AWS S3. ({issue}`18976`)

Reviewer: Improve performance of ...

Author: I think this is more accurate - there was a performance regression, and this fixes it.

## Iceberg connector

* Add support for usage of `date` and `timestamp` arguments in `FOR TIMESTAMP AS
  OF` expressions. ({issue}`14214`)
* Reduce coordinator memory usage when running queries on tables with large
  metadata. ({issue}`19009`)

## Kudu connector

> * Add support for pushing down `UPDATE` statements to Kudu. ({issue}`16445`)

Reviewer: Did we always say "to x"

## MySQL connector

* Add support for pushing down `UPDATE` statements to MySQL. ({issue}`16445`)

## Oracle connector

* Add support for pushing down `UPDATE` statements to Oracle. ({issue}`16445`)
* Fix potential query failure when joins are pushed down to Oracle. ({issue}`18924`)

## Phoenix connector

* Add support for pushing down `UPDATE` statements to Phoenix. ({issue}`16445`)

## PostgreSQL connector

* Add support for pushing down `UPDATE` statements to PostgreSQL. ({issue}`16445`)

## Redshift connector

* Add support for pushing down `UPDATE` statements to Redshift. ({issue}`16445`)

> 

Reviewer 3: @vlad-lyutenko Ignte and SingleStore connector don't support UPDATE statement in #16445?

Reviewer 4: Supported connectors:

Ignite
MariaDB
MySql
Oracle
Postgres
Redshift
Singlestore
SqlServer
